include/RAJA/policy/cuda/forall.hpp

```diff
@@ -197,39 +197,31 @@ RAJA_INLINE void forall(cuda_exec<BLOCK_SIZE, Async>,
                         Iterable&& iter,
                         LOOP_BODY&& loop_body)
 {
-  beforeCudaKernelLaunch();
+  auto begin = std::begin(iter);
+  auto end   = std::end(iter);
 
-  auto body = loop_body;
+  auto len = std::distance(begin, end);
 
-  auto first_begin = std::begin(iter);
-  auto final_end = std::end(iter);
-  auto total_len = std::distance(first_begin, final_end);
-  auto max_step_size = (getCudaMemblockUsedCount() > 0)
-                                 ? BLOCK_SIZE * RAJA_CUDA_MAX_NUM_BLOCKS
-                                 : total_len;
+  if (len > 0 && BLOCK_SIZE > 0) {
 
-  RAJA_FT_BEGIN;
+    auto gridSize = RAJA_DIVIDE_CEILING_INT(len, BLOCK_SIZE);
 
-  for (decltype(total_len) step_size, offset = 0; offset < total_len;
-       offset += step_size) {
+    RAJA_FT_BEGIN;
 
-    step_size = RAJA_MIN(total_len - offset, max_step_size);
+    cudaStream_t stream = 0;
 
-    auto begin = first_begin + offset;
-    auto end = begin + step_size;
+    forall_cuda_kernel<<<gridSize, BLOCK_SIZE, 0, stream>>>(
+        RAJA::cuda::createLaunchBody(
+            gridSize, BLOCK_SIZE, 0, stream,
+            std::forward<LOOP_BODY>(loop_body)),
+        std::move(begin), len);
+    cudaErrchk(cudaPeekAtLastError());
```

Comment:
I noticed the abstraction of cuda synchronize. Is it possible to abstract away error checking into a clean API call as well?